### PR TITLE
[SD-876] Only restart api server when port number changed

### PR DESCRIPTION
**Pull request by wemrysi**
Changes the behavior of `FileSystemApi` to only restart the api server when the port number is changed. For mounting changes, a local cache of the configuration is updated (in a `TaskRef`) after writing the updated config to disk.

The server main is also refactored in terms of `scalaz.stream.Process` which makes running + restarting uniform and allows us to test restarting the server deterministically.

### CODE REVIEW

File: web/src/main/scala/slamdata/engine/api/fs.scala [hunk 1]
```diff
@@ -343,73 +357,78 @@ final case class FileSystemApi(backend: Backend, contentPath: String, config: Co
...
-            else
-              -\/(Conflict(errorBody("Can't add a mount point above the existing mount point at " + k, None)))))
-        }.sequenceU.fold(
-          ɩ,
-          κ(respond(for {
-            upd <- addPath(path, req)
-          } yield (if (upd) "updated" else "added") + " " + path)))
+        ref.read.flatMap(cfg =>
+          (if (cfg.mountings contains path) None else ensureNoOverlaps(cfg, path)) getOrElse
+          respond(addPath(cfg, path, req) map (upd => s"${upd ? "updated" | "added"} $path")))
```
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: Although we have no explicit policy, we've been avoiding string interpolation almost entirely, and certainly not putting logic in it. I know @jdegoes is not a fan; good time to make a decision and maybe add something to the wiki?

File: web/src/main/scala/slamdata/engine/api/server.scala [hunk 2]
```diff
@@ -27,11 +27,17 @@ import scala.concurrent.duration._
...
-import scalaz.effect._
+import scalaz.stream._
+
+import org.http4s.server.{Server => Http4sServer}
+import org.http4s.server.blaze.BlazeBuilder
 
 object Server {
-  private val serv =
-    IO.newIORef(None: Option[org.http4s.server.Server]).unsafePerformIO
+  private def info(msg: => String): Task[Unit] =
```
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/server.scala]: A good idea but can we find a better home for these?

File: web/src/main/scala/slamdata/engine/api/server.scala [hunk 3]
```diff
@@ -48,51 +54,85 @@ object Server {
...
+      case -\/(err)                         => Task.fail(err)
+      case \/-(s)                           => Task.delay(s.close()).as(None)
+    })
 
-    config => for {
-      _       <- configWriter(config)
-      _       <- runAsync(restart(config).run.map(ignore))
-    } yield ()
+  /** An available port number. */
+  def anyAvailablePort: Task[Int] = Task.delay {
```
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/server.scala]: Code moving around makes it hard to follow this diff. Which I'm ok with if there's a good reason to re-order things. Was there?

File: web/src/main/scala/slamdata/engine/api/fs.scala [hunk 4]
```diff
@@ -317,21 +325,27 @@ final case class FileSystemApi(backend: Backend, contentPath: String, config: Co
...
+              //       on a random one, thus this response text may not be accurate.
+              resp <- Ok("changed port to " + i)
+            } yield resp))
+
       case DELETE -> Root / "port" => for {
-        _    <- reloader(config.copy(server = SDServerConfig(None)))
-        resp <- Ok("reverted to default port")
+        cfg  <- ref.read
+        dprt =  SDServerConfig.DefaultPort
+        _    <- reloader(cfg.copy(server = SDServerConfig(Some(dprt))))
```
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: I think we want `None` here, just to avoid writing the (current) default port to the config file. If we ever change our minds on it, we don't want to have to ask users to go in and manually delete it.

File: web/src/main/scala/slamdata/engine/api/fs.scala [hunk 5]
```diff
@@ -343,73 +357,78 @@ final case class FileSystemApi(backend: Backend, contentPath: String, config: Co
...
-        if (config.mountings.contains(path))
-          for {
-            _    <- reloader(config.copy(mountings = config.mountings - path))
-            resp <- Ok("deleted " + path)
-          } yield resp
-        else
-          Ok()
+        ref.read.flatMap(cfg =>
+          if (cfg.mountings contains path)
+            writeConfig(ref, cfg.copy(mountings = cfg.mountings - path)) *>
```
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: Not sure I like the trailing `*>` as an alternative to `for { _ <- ...`, but this is another thing I feel deserves to be approached consistently.

File: web/src/test/scala/slamdata/engine/api/fs.scala [hunk 6]
```diff
@@ -1697,62 +1714,76 @@ class ApiSpecs extends Specification with DisjunctionMatchers with PendingWithAc
...
     }
+  }
 
-    "restart sequence" should {
-      // Exercises the server restarting code path in server.scala.
-       def withRestartableServer[A](backend: Backend, config: Config)(body: => A): A = {
-         val rez: slamdata.engine.Evaluator.EnvironmentError \/ Int = Server.run(port, config, "", 1.seconds, tester, mounter(backend), configWriter).run.run
+  "/server" should {
+    def withServerExpectingRestart[A, B](backend: Backend, config: Config, timeoutMillis: Long = 10000)
+                                        (causeRestart: => A)(afterRestart: => B): B = {
```
Comment (mossprescott) [inline web/src/test/scala/slamdata/engine/api/fs.scala]: :grinning: I like it.

File: web/src/main/scala/slamdata/engine/api/fs.scala (hunk 5 above)
Comment (jdegoes) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: The reason for `*>` is that it's strictly less powerful than `>>=` (AKA `flatMap`). It's considered good practice (one I agree with) to use the least specific abstraction that gets the job done. 

However, I think we do need some standardization on `flatMap` versus `for { ... }` since we have mixed styles; also on various syntactical alternatives to `ap` (`|@|`, etc.).

File: web/src/main/scala/slamdata/engine/api/server.scala (hunk 2 above)
Comment (jdegoes) [inline web/src/main/scala/slamdata/engine/api/server.scala]: Also probably worth pulling in [Journal](https://github.com/oncue/journal) or similar at some point.

File: web/src/main/scala/slamdata/engine/api/fs.scala (hunk 4 above)
Comment (wemrysi) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: Ah, you're absolutely right, that's much better behavior (and indeed what was there originally!).

File: web/src/main/scala/slamdata/engine/api/fs.scala (hunk 5 above)
Comment (wemrysi) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: Agree on the need to standardize, I tend (and prefer) to use combinators where possible out of habit for the reason John mentions. I also think they are clearer in many cases, i.e.

```
for {
  a <- fa
} yield g(a)
```
vs `fa map g`, or

```
for {
  a <- fa
  b <- g(a)
} yield b
```
vs `fa flatMap g` or even `fa >>= g`.

That said, I'll try not to change styles just for change's sake, only when it is a strict improvement.

File: web/src/main/scala/slamdata/engine/api/fs.scala (hunk 1 above)
Comment (wemrysi) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: Ok, thanks for pointing that out, I'll omit the interpolation. This probably isn't the best place to discuss, but what are the arguments against it?

File: web/src/main/scala/slamdata/engine/api/server.scala (hunk 3 above)
Comment (wemrysi) [inline web/src/main/scala/slamdata/engine/api/server.scala]: The only reason was to keep the file organized in a more-or-less top-down fashion, but I sympathize with moving code making things hard to understand. I should have moved things in one commit and modified them in another to make things more clear.

File: web/src/main/scala/slamdata/engine/api/fs.scala (hunk 1 above)
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: I can't speak for the rest. My own preference would probably be to use simple interpolation but avoid anything more complicated. Good topic for a tech hangout?

File: web/src/main/scala/slamdata/engine/api/fs.scala [hunk 7]
```diff
@@ -124,13 +124,21 @@ object ResponseFormat {
   }
 }
 
-final case class FileSystemApi(backend: Backend, contentPath: String, config: Config, tester: BackendConfig => EnvTask[Unit], reloader: Config => Task[Unit]) {
+final case class FileSystemApi(
+  backend: Backend, contentPath: String, initialConfig: Config,
+  tester: BackendConfig => EnvTask[Unit],
+  reloader: Config => Task[Unit],
```
Comment (jedesah) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: Sorry, a little late to the party, what is the difference between `reloader` and `configWriter`? I think it would not be unreasonable to add a little bit of documentation on this class to explain the various fields.

Comment (wemrysi) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: Yeah, the types here aren't very helpful =(. In short, `reloader` starts a new server with the provided `Config` whilst `configWriter` persists the given `Config`. Consider this a +1 for tracking effects in the types!

Comment (jedesah) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: Not sure I understand, aren't `reloader` and `Config` kind of like callbacks essentially? I don't understand what you mean by
>starts a new server with the provided Config

I really think we should add some ScalaDoc here. I will be glad to do it once I have a better understanding of what the different fields are.
